VMware dvPG description and uplink failover order features

traittinen opened the pull request:
##### SUMMARY
Added two new features. Ability to:
- Add description to a portgroup
- Change portgroup uplink failover order

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
vmware_dvs_portgroup

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.0
  config file = None
  configured module search path = [u'/home/tomi/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Nov 20 2015, 02:00:19) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->
With existing vmware_dvs_portgroup module creating a new portgroup always inherits the uplink failover settings. These needs to be changed e.g. in cases where LAG is used as an uplink since standard Uplinks are always inherited by default.

<!--- Paste verbatim command output below, e.g. before and after your change -->
Example plays for usage:
```
    - name: Create vlan distributed portgroup 112
      connection: local
      vmware_dvs_portgroup:
        hostname: "{{ vcenter_ip }}"
        username: "{{ vcenter_user }}"
        password: "{{ vcenter_pw }}"
        validate_certs: false
        portgroup_name: test-pg-112
        portgroup_description: Test portgroup VLAN 112
        switch_name: Production01
        vlan_id: 112
        num_ports: 120
        portgroup_type: ephemeral
        teaming_policy:
          active_uplinkport: lag1
        state: present

    - name: Create vlan distributed portgroup 113
      connection: local
      vmware_dvs_portgroup:
        hostname: "{{ vcenter_ip }}"
        username: "{{ vcenter_user }}"
        password: "{{ vcenter_pw }}"
        validate_certs: false
        portgroup_name: test-pg-113
        portgroup_description: Test portgroup VLAN 113
        switch_name: Production01
        vlan_id: 113
        num_ports: 120
        portgroup_type: ephemeral
        teaming_policy:
          active_uplinkport:
            - Uplink 1
            - Uplink 2
        state: present
```

traittinen committed:
DVS portgroup description and uplink port order

Added two new features:
- Ability to add description to a portgroup
- Ability to change uplink port failover order

traittinen committed:
Deleted unnecessary quote

traittinen commented:
@Akasurde @dav1x @pdellaert @warthog9 can anyone help out here? Can't figure out why the test fails..

Akasurde commented:
@traittinen On line 112, you need to add default values for `active_uplinkport` and `standby_uplinkport` Let me know if you need any help on that.

traittinen commented:
Thanks @Akasurde! All checks pass now.

Akasurde commented:
@traittinen Could you please rebase branch and let me know ? Thanks.

traittinen commented:
@Akasurde I think this is now ok.

pdellaert commented on lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup.py:
```diff
@@ -96,13 +101,17 @@
             - '- C(inbound_policy) (bool): Indicate whether or not the teaming policy is applied to inbound frames as well. (default: False)'
             - '- C(notify_switches) (bool): Indicate whether or not to notify the physical switch if a link fails. (default: True)'
             - '- C(rolling_order) (bool): Indicate whether or not to use a rolling policy when restoring links. (default: False)'
+            - '- C(active_uplinkport) (list): List of active uplink ports used for load balancing. (default: [])'
```
Please add to the description that if `active_uplinkport` is not set (or set to an empty list), `standby_uplinkport` is ignored, even if it is set.

pdellaert reviewed (changes requested):
Overall good PR, few requests for changes.

Please add an example of how to use this for:
* one active and one standby uplink
* two active uplinks
* portgroup description (this can be combined with any of the others)

Also have a look at adding a test for this?

traittinen commented:
Accidentally deleted branch patch-1. Reopening PR..

ckotte commented:
@traittinen Could we do it like this instead?

```
- name: Add Portgroup
  vmware_dvs_portgroup:
    hostname: '{{ vcenter_hostname }}'
    username: '{{ vcenter_username }}'
    password: '{{ vcenter_password }}'
    switch: dvSwitch
    portgroup_name: Test
    teaming:
        load_balancing: failover_explicit
        network_failure_detection: link_status_only
        notify_switches: true
        failback: true
        failover_order:
            active_nics:
                - lag1
  delegate_to: localhost
```

This would match with the naming in the PR https://github.com/ansible/ansible/pull/47052 for the vmware_portgroup module.

ckotte commented on lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup.py:
```diff
@@ -96,13 +101,17 @@
             - '- C(inbound_policy) (bool): Indicate whether or not the teaming policy is applied to inbound frames as well. (default: False)'
             - '- C(notify_switches) (bool): Indicate whether or not to notify the physical switch if a link fails. (default: True)'
             - '- C(rolling_order) (bool): Indicate whether or not to use a rolling policy when restoring links. (default: False)'
+            - '- C(active_uplinkport) (list): List of active uplink ports used for load balancing. Default uplink is used if not defined. (default: [])'
```
Can you change this to `active_uplinks` and `standby_uplinks`, please?

ckotte commented on lib/ansible/modules/cloud/vmware/vmware_dvs_portgroup.py:
```diff
@@ -35,6 +35,11 @@
         description:
             - The name of the portgroup that is to be created or deleted.
         required: True
+    portgroup_description:
```
Can you change this to just `description` and an alias of `portgroup_description` if you want.

traittinen committed:
Adding changes to changed base.

traittinen committed:
Changed following attributes:
portgroup_description -> description
active_uplinkport -> active_uplinks
standby_uplinkport -> standby_uplinks

traittinen committed:
Sanity check fixing.

traittinen committed:
Fixed documentation default values.

traittinen committed:
YAML syntax error fixes.

traittinen committed:
Syntax fixing.

traittinen commented:
@ckotte this should ok now. Had a little problems with rebasing which caused little delays..

ckotte commented:
> @ckotte this should be ok now. Had a little problems with rebasing which caused little delays..

Ok. I implemented the same in my refactoring PR #48719 with many other improvements. 

Did you check the behavior of `default=[]`? This didn't work on my side and I had to add if statements to set the uplinks to an empty array. This basically is only important if you do the update since you need to check if the current uplinks (array) match the array specified in the options:
```
active_uplinks = pg_config.defaultPortConfig.uplinkTeamingPolicy.uplinkPortOrder.activeUplinkPort
if active_uplinks != self.teaming_failover_order_active:
  [...]
```

Maybe if you have time you can check this PR.

traittinen commented:
> Did you check the behavior of `default=[]`? This didn't work on my side and I had to add if statements to set the uplinks to an empty array.

Yes, it seems to work. New pg inherits default uplinks as active uplinks if the array is empty/not defined.

> This basically is only important if you do the update since you need to check if the current uplinks (array) match the array specified in the options

Correct me if I'm wrong, but updating pg is not implemented currently. If the matching pg name is found from vCenter/dvs the configuration is not changed.

traittinen commented:
> Really? If the array is empty, then no uplinks are configured. `None` will be changed to `[]` as well. I had to set the uplinks explicitly to get the default behavior...

Yep. Just tested this. If active_uplink is not configured the defaults are used.
